core/server/master/src/main/java/alluxio/underfs/UfsStatusCache.java

```diff
@@ -137,9 +142,11 @@ public Collection<UfsStatus> addChildren(AlluxioURI path, Collection<UfsStatus>
   public UfsStatus remove(AlluxioURI path) {
     Preconditions.checkNotNull(path, "can't remove null status cache path");
     UfsStatus removed = mStatuses.remove(path);
+    if (removed != null) {
+      // Update global counters for all InodeSyncStream
+      DefaultFileSystemMaster.Metrics.UFS_STATUS_CACHE_SIZE_TOTAL.dec();
+    }
     int childrenCnt = mChildren.containsKey(path) ? mChildren.get(path).size() : 0;
-    // Update global counters for all InodeSyncStream
-    DefaultFileSystemMaster.Metrics.UFS_STATUS_CACHE_SIZE_TOTAL.dec();
     DefaultFileSystemMaster.Metrics.UFS_STATUS_CACHE_CHILDREN_SIZE_TOTAL.dec(childrenCnt);
     mChildren.remove(path); // ok if there aren't any children
```

Comment:
@jiacheliu3  If there are any children, do we need to remove them from `mStatuses`, too? Since children are added to `mStatuses` in `addChildren`.

Reply:
Very good point. I think it's because of how the InodeSyncStream uses this cache. InodeSyncStream will process each path roughly following BFS, thus parent first then each child.  At the end of processing the path will be removed from the status cache
https://github.com/Alluxio/alluxio/blob/92c5400b29ccdf064940ee943371a0e7746133bf/core/server/master/src/main/java/alluxio/master/file/InodeSyncStream.java#L519

This suggests we cannot remove each child status from the big `mStatuses` map because they will be used when the children paths are traversed. And the current code works.

How much I don't like the cohesion is another topic...

Reply:
Then the consistency between the two counters has to depend on `InodeSyncStream`s, if I understand this correctly.

Reply:
yes, depends on the InodeSyncStream traversing in the correct order